src/lib.rs

```diff
@@ -43,6 +67,38 @@ mod output_template {
     }
 }
 
+/// Create a structured TypstError from diagnostics
+fn create_typst_error(
+    py: Python<'_>,
+    world: &SystemWorld,
+    errors: &[SourceDiagnostic],
+    warnings: &[SourceDiagnostic],
+) -> PyResult<PyErr> {
+    // Get the main error message by formatting all diagnostics
+    let formatted_message = crate::compiler::format_diagnostics(world, errors, warnings)
+        .unwrap_or_else(|_| "Failed to format diagnostic message".to_string());
+
+    // Extract structured information from the first error (most relevant)
+    let primary_error = errors.first();
+    
+    let (hints, trace) = if let Some(error) = primary_error {
+        // Extract hints
+        let hints = error.hints.iter().map(|h| h.to_string()).collect::<Vec<_>>();
+        
+        // Extract trace information
+        let trace = error.trace.iter().map(|point| {
+            format!("at {}", point.v)
+        }).collect::<Vec<_>>();
+        
+        (hints, trace)
+    } else {
+        (Vec::new(), Vec::new())
+    };
+
+    // Create the TypstError with structured details
+    TypstError::new_with_details(py, formatted_message, hints, trace)
```

Comment:
Let's define an intermediate Rust structure to hold error details and only convert to `TypstError` after calling `compile`/`query` in `allow_threads` so we only hold GIL when converting the error type.